src/sys/freebsd.rs

```diff
@@ -1 +1,268 @@
+        Err(io::Error::last_os_error())
+    } else {
+        Ok(())
+    }
+}
+
+pub fn list_fd(fd: RawFd) -> io::Result<XAttrs> {
+    let sysvec = unsafe {
+        let res = allocate_loop(|buf, len| extattr_list_fd(fd, EXTATTR_NAMESPACE_SYSTEM, buf as *mut c_void, len as size_t));
+        // On FreeBSD, system attributes require root privileges to view. However,
+        // to mimic the behavior of listxattr in linux and osx, we need to query
+        // them anyway and return empty results if we get EPERM
+        match res {
+            Ok(v) => v,
+            Err(err) => {
+                if err.raw_os_error().unwrap() == EPERM {
+                    Vec::new()
+                } else {
+                    return Err(err)
+                }
+            }
+        }
+    };
+
+    let uservec = unsafe {
+        let res = allocate_loop(|buf, len| extattr_list_fd(fd, EXTATTR_NAMESPACE_USER, buf as *mut c_void, len as size_t));
+        match res {
+            Ok(v) => v,
+            Err(err) => {
+                return Err(err)
+            }
+        }
+    };
+
+    Ok(XAttrs {
+        system_attrs: sysvec.into_boxed_slice(),
+        user_attrs: uservec.into_boxed_slice(),
+        offset: 0,
+    })
+}
+
+pub fn get_path(path: &Path, name: &OsStr) -> io::Result<Vec<u8>> {
+    let (ns, name) = try!(name_to_ns(name));
+    let path = try!(path_to_c(path));
+    unsafe {
+        allocate_loop(|buf, len| extattr_get_link(path.as_ptr(), ns, name.as_ptr(), buf as *mut c_void, len as size_t))
+    }
+}
+
+pub fn set_path(path: &Path, name: &OsStr, value: &[u8]) -> io::Result<()> {
+    let (ns, name) = try!(name_to_ns(name));
+    let path = try!(path_to_c(path));
+    let ret = unsafe {
+        extattr_set_link(
+            path.as_ptr(),
+            ns,
+            name.as_ptr(),
+            value.as_ptr() as *const c_void,
+            value.len() as size_t)
+    };
+    if ret == -1 {
+        Err(io::Error::last_os_error())
+    } else {
+        Ok(())
+    }
+}
+
+pub fn remove_path(path: &Path, name: &OsStr) -> io::Result<()> {
+    let (ns, name) = try!(name_to_ns(name));
+    let path = try!(path_to_c(path));
+    let ret = unsafe { extattr_delete_link(path.as_ptr(), ns, name.as_ptr()) };
+    if ret != 0 {
+        Err(io::Error::last_os_error())
+    } else {
+        Ok(())
+    }
+}
+
+pub fn list_path(path: &Path) -> io::Result<XAttrs> {
+    let path = try!(path_to_c(path));
+    let sysvec = unsafe {
+        let res = allocate_loop(|buf, len| extattr_list_link(path.as_ptr(), EXTATTR_NAMESPACE_SYSTEM, buf as *mut c_void, len as size_t));
+        // On FreeBSD, system attributes require root privileges to view. However,
+        // to mimic the behavior of listxattr in linux and osx, we need to query
+        // them anyway and return empty results if we get EPERM
+        match res {
+            Ok(v) => v,
+            Err(err) => {
+                if err.raw_os_error().unwrap() == EPERM {
```

Comment:
Personally, I don't like using unwrap unless I absolutely need to. A safer way to do this is `err.raw_os_error() == Some(EPERM)`.